<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Запрос-к-базе-данных" data-toc-modified-id="Запрос-к-базе-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Запрос к базе данных</a></span></li><li><span><a href="#SQL-запросы" data-toc-modified-id="SQL-запросы-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>SQL-запросы</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

# Анализ базы данных сервиса для чтения книг по подписке

Эпидемия коронавируса изменила привычный порядок вещей на некоторое время. У жителей городов появилось много свободного времени, которое нужно провести в уединении. Чтение книг один из вариантов такого досуга. Количество сервисов по чтению книг по подписке стало расти. 

Для одного из них необходимо провести анализ базы данных. В ней  вся информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. На основе анализа этих данных будет сформулированы предложения для создания нового продукта.

Данные в базе находятся в таблицах
 - books
 - authors
 - publishers
 - ratings
 - reviews


**Задачи данной работы:**

 - выгрузить информацию из базы данных в файлы `csv`
 
 - исследовать таблицы
 
 - написать SQL-запросы
 

**Таким образом работа пройдет по следующим этапам:**

 - создадим коннекцию к базе
 - напишем код для выгрузки данных из таблицы в csv-файлы
 - изучим данные
 - напишем SQL-запросы
  

## Запрос к базе данных

 - Создадим запрос к базе данных и сохраним информацию в csv-файл

In [1]:
#!pip install ipython-sql               # импортируем библиотеки
#!pip3 install psycopg2

import pandas as pd
from sqlalchemy import create_engine  


In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                        db_config['pwd'],
                                                        db_config['host'],
                                                        db_config['port'],
                                                        db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
                                               

In [3]:
query = '''                  
           SELECT *
           FROM books
        '''

books_raw = pd.io.sql.read_sql(query, con = engine)
books_raw.to_csv('books.csv', index = False)
books = books_raw
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


In [5]:
books['book_id'].nunique()

1000

In [6]:
query = '''                  
           SELECT *
           FROM authors
        '''

authors_raw = pd.io.sql.read_sql(query, con = engine)
authors_raw.to_csv('authors.csv', index = False)
authors =authors_raw
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [7]:
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


In [8]:
query = '''                  
           SELECT *
           FROM publishers
        '''

publishers_raw = pd.io.sql.read_sql(query, con = engine)
publishers_raw.to_csv('publishers.csv', index = False)
publishers =publishers_raw
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [9]:
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


In [10]:
query = '''                  
           SELECT *
           FROM ratings
        '''

ratings_raw = pd.io.sql.read_sql(query, con = engine)
ratings_raw.to_csv('ratings.csv', index = False)
ratings = ratings_raw
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [11]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


In [12]:
query = '''                  
           SELECT *
           FROM reviews
        '''

reviews_raw = pd.io.sql.read_sql(query, con = engine)
reviews_raw.to_csv('reviews.csv', index = False)
reviews = reviews_raw
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [13]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


In [14]:
tables = [books, authors, publishers, ratings, reviews]
for i in tables:
    print(i.duplicated().sum())

0
0
0
0
0


**Вывод**

Данные представлены в корректном виде
 - пропусков нет
 - дубликатов нет
 - типы танных соответствуют ожидаемым. Исключения таблица books, где дата записана как тип object, но в самом sql-запросе мы сделали преобразование

## SQL-запросы

 - *Посчитайте,сколько книг вышло после 1 января 2000 года*

In [15]:
query_1 = ''' SELECT COUNT(book_id)
               FROM books 
               WHERE publication_date::date >= '2000-01-01'
               '''

In [16]:
result_1 = pd.read_sql(query_1, con = engine)
print(result_1)

   count
0    821


Комментарий

- Начиная с 1 января 2000 года вышло 821 книга

 - *Для каждой книги посчитайте количество обзоров и среднюю оценку*

In [17]:
query_2 = '''SELECT b.book_id,
                    COUNT(rv.review_id) AS review_count,
                    ROUND(AVG(rt.rating),1) AS rating_avg
             FROM books b
             LEFT JOIN ratings rt ON b.book_id = rt.book_id
             LEFT JOIN reviews rv ON b.book_id = rv.book_id
             GROUP BY b.book_id
             
             
          '''

In [18]:
result_2 = pd.read_sql(query_2, con = engine)
print(result_2)

     book_id  review_count  rating_avg
0        652             4         4.5
1        273             4         4.5
2         51            60         4.3
3        951             4         4.0
4        839            28         4.3
..       ...           ...         ...
995       64            52         4.2
996       55             4         5.0
997      148            21         3.4
998      790             4         3.5
999      828             4         3.0

[1000 rows x 3 columns]


Комментарий

 - Количество обзоров и среднюю оценку для каждой книги посчитали


- *Определите издательство, которое выпустило наибольшее число книг толще 50 страниц*

In [19]:
query_3 = '''  SELECT p.publisher,
               COUNT(b.book_id) AS num
                    
             FROM publishers p
             JOIN books b ON b.publisher_id = p.publisher_id
             WHERE b.num_pages > 50
             GROUP BY p.publisher
             ORDER BY num DESC
             LIMIT 1
                    
             '''

In [20]:
result_3 = pd.read_sql(query_3, con = engine)
print(result_3)

       publisher  num
0  Penguin Books   42


Комментарий
 - Penguin Books - издательство, которое выпустило больше всего книг, толще 50 страниц. Таких книг у него 42

 - *Определите автора с самой высокой средней оценкой книг. Учитывайте только книги с 50 и более оценками*

In [21]:
query_4 = '''WITH h AS(SELECT a.author,
                    COUNT(r.rating_id) AS counter,
                    ROUND(AVG(r.rating), 2) AS rating
                    FROM books b
                    JOIN authors a ON b.author_id = a.author_id
                    LEFT JOIN ratings r ON b.book_id = r.book_id
                    GROUP BY a.author)
             SELECT h.author,
                    h.rating
            FROM h
            WHERE h.counter >= 50
            GROUP BY 1, 2
            ORDER BY 2 DESC
            LIMIT 1
                      
          '''

In [22]:
result_4 = pd.read_sql(query_4, con = engine)
print(result_4)

           author  rating
0  Diana Gabaldon     4.3


Комментарий
 - Диана Гэблдон автор автор с самой высокой средней оценкой книг (для книг с 50 оценками и более)

 - Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [23]:
query_5 = '''WITH r AS (SELECT 
                    rv.username,
                    COUNT(rt.rating_id) AS rat,
                    COUNT(rv.review_id) AS rev
                FROM ratings rt
                JOIN reviews rv ON rt.username = rv.username
                GROUP BY 1)
                SELECT ROUND(AVG(r.rev))  AS avg_review
                 FROM r     
                WHERE r.rat >50
                
            '''

In [24]:
result_5 = pd.read_sql(query_5, con = engine)
print(result_5)

   avg_review
0       720.0


Комментарий

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок  - 720

## Вывод


Мы создали коннекцию к база данных и выгрузили таблицы

 - books
 - authors
 - publishers
 - ratings
 - reviews
 
Изучили данные, проверили на дубликаты 
 
Написали SQL- запросы для ответов на вопросы задания и получили следующую информацию

 - После 1 января 2000 года вышло 821 книга
 - Посчитали количество обзоров и среднюю оценку для каждой книги
 - Penguin Books издательство, которое выпустило больше всего книг толще 50 страниц. Всего таких книг - 42
 - Диана Гэблдон - автор книг с самой высокой средней оценкой книг  (для книг с 50 оценками и более)
 - Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 720 